onnx_caffe2/backend.py

```diff
@@ -297,7 +299,65 @@ def _create_gemm(cls, n, opset_version):
         return ops
 
     @classmethod
-    def _create_pad(cls, n, opset_version):
+    def _create_lstm(cls, onnx_model, model_helper, n, opset_version):
+        assert model_helper is not None, "cannot convert LSTMs through this codepath"
+        assert onnx_model is not None, "cannot convert LSTMs through this codepath"
+        allowed_attrs = {'hidden_size'}
+        for x in n.attrs.keys():
+            assert x in allowed_attrs, "unsupported attribute " + x + " to LSTM"
+
+        input_blob, W, R, B, sequence_lens, initial_h, initial_c = n.inputs
+
+        hidden_size = n.attrs['hidden_size']
+
+        # ad-hoc, informally-specified, bug-ridden, slow
+        # implementation of shape inference
+        for x in onnx_model.graph.input:
+            if x.name == input_blob:
+                input_size = x.type.tensor_type.shape.dim[2].dim_value
+        for x in map(OnnxNode, onnx_model.graph.node):
+            if x.op_type == n.op_type and x.outputs[0] == input_blob:
+                input_size = x.attrs['hidden_size']
+
+        name = dummy_name()
+
+        # input and recurrence biases are squashed together in onnx but not in caffe2
+        model_helper.net.Slice(B, name + "_bias_i2h",   starts=[0],             ends=[4*hidden_size])
+        model_helper.net.Slice(B, name + "_bias_gates", starts=[4*hidden_size], ends=[8*hidden_size])
+
+        # caffe2 has a different order from onnx. We need to rearrange
+        #   i o f c -> i f o c
+        def reform(name_from, name_to, extra_starts, extra_ends):
+            for start, suffix in enumerate(("_i", "_o", "_f", "_c")):
+                model_helper.net.Slice(
+                    name_from,
+                    name_from + suffix,
+                    starts = [   start  * hidden_size] + extra_starts,
+                    ends =   [(1+start) * hidden_size] + extra_ends)
+            model_helper.net.Concat(
+                [name_from + suffix for suffix in ("_i", "_f", "_o", "_c")],
+                [name_to, dummy_name()], axis=0)
+        reform(W, name + "/i2h_w", [0], [input_size])
+        reform(R, name + "/gates_t_w", [0], [hidden_size])
+        reform(name + "_bias_i2h", name + "/i2h_b", [], [])
+        reform(name + "_bias_gates", name + "/gates_t_b", [], [])
+
+        lstm_out = rnn_cell.LSTM(
+            model_helper,
+            input_blob,
+            sequence_lens,
+            [initial_h, initial_c],
+            input_size,
+            hidden_size,
+            name,
+            forward_only=True,
+        )
+        hidden_t_all, hidden_t_last, _, _ = lstm_out
+        model_helper.net.Copy(hidden_t_all, n.outputs[0])
```

Comment:
this is somewhat hacky. You can 'rename' the blobs in the net (e.g. with net.Clone). Or at least add a TODO for it

Reply:
yeah, this is part of dealing with the particular interface of rnn_cell.LSTM, I have no control over the output names.

I might try out Clone, though there's some annoying details there as well (do I Clone every time I process an LSTM op, or try to collect information and do it once at the end, etc.)